### PR TITLE
test: increase test coverage in internal/kafka/internal/clusters

### PR DESCRIPTION
**Pull request by JasonPowr**
## Description

Increase unit test code coverage in internal/kafka/internal/clusters where possible or ignore files, for which writing unit tests is not feasible in the make test make target.

Jira: [Here](https://issues.redhat.com/browse/MGDSTRM-7975)

Test coverage:
|  File  | Test Coverage Before  | Test coverage after | 
|---|---|---|
| cluster_builder.go |  100% | 100%  |
| ocm_provider.go |  87.4% | 87.4%  |
| provider.go |  0% | 100%  |
| standalone_provider.go |  34.4% | 47.7%  |
| types.go |  100% | 100%  |

## Verification Steps
1. Run `make test`.
2. Run `make test/html/coverage/report`.
3. Can see the new lines added in the report.

## Checklist (Definition of Done)
- [x] All acceptance criteria specified in JIRA have been completed
- [x] Unit and integration tests added that prove the fix is effective or the feature works (tested against emulated and non-emulated OCM environment)
- [ ] ~~Documentation added for the feature~~
- [ ] CI and all relevant tests are passing
- [ ] Code Review completed
- [ ] Verified independently by reviewer
- [ ] ~~Required metrics/dashboards/alerts have been added (or PR created).~~
- [ ] ~~Required Standard Operating Procedure (SOP) is added.~~
- [ ] ~~JIRA has created for changes required on the client side~~

### CODE REVIEW

File: internal/kafka/internal/clusters/provider_test.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+					},
+				},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := NewDefaultProviderFactory(tt.args.ocmClient, tt.args.connectionFactory, tt.args.ocmConfig, tt.args.awsConfig, tt.args.dataplaneClusterConfig); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("NewDefaultProviderFactory() = %v, want %v", got, tt.want)
+			}
```
Comment (ziccardi) [inline internal/kafka/internal/clusters/provider_test.go:45]: I find this hard to read. I think would be more readable to have it like this:
```suggestion
			got := NewDefaultProviderFactory(tt.args.ocmClient, tt.args.connectionFactory, tt.args.ocmConfig, tt.args.awsConfig, tt.args.dataplaneClusterConfig)
			Expect(got).To(Equal(tt.want))
```

File: internal/kafka/internal/clusters/provider_test.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			d := &DefaultProviderFactory{
+				providerContainer: tt.fields.providerContainer,
+			}
+			got, err := d.GetProvider(tt.args.providerType)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("DefaultProviderFactory.GetProvider() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
```
Comment (ziccardi) [inline internal/kafka/internal/clusters/provider_test.go:114]:
```suggestion
Expect(err!=nil).To(Equal(tt.wantErr))
```

File: internal/kafka/internal/clusters/provider_test.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("DefaultProviderFactory.GetProvider() = %v, want %v", got, tt.want)
+			}
```
Comment (ziccardi) [inline internal/kafka/internal/clusters/provider_test.go:117]:
```suggestion
			Expect(got).To(Equal(tt.want))
```

File: internal/kafka/internal/clusters/standalone_provider_test.go [hunk 4]
```diff
@@ -1046,3 +1040,202 @@ func TestStandaloneProvider_buildKasFleetshardOperatorSubscription(t *testing.T)
...
+			},
+		},
+	}
+	RegisterTestingT(t)
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			provider := newStandaloneProvider(test.fields.connectionFactory, test.fields.dataplaneClusterConfig)
+			ok, err := provider.InstallKasFleetshard(test.args.clusterSpec, test.args.params)
+			Expect(err != nil).To(Equal(test.wantErr))
+			Expect(ok).To(Equal(true))
```
Comment (ziccardi) [inline internal/kafka/internal/clusters/standalone_provider_test.go:1118]:
```suggestion
			Expect(ok).To(BeTrue())
```

File: internal/kafka/internal/clusters/standalone_provider_test.go [hunk 5]
```diff
@@ -1046,3 +1040,202 @@ func TestStandaloneProvider_buildKasFleetshardOperatorSubscription(t *testing.T)
...
+				newConfiguration: "lastAppliedConfigurationAnnotation",
+			},
+			want: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := shouldApplyChanges(tt.args.dynamicClient, tt.args.existingObj, tt.args.newConfiguration); got != tt.want {
+				t.Errorf("shouldApplyChanges() = %v, want %v", got, tt.want)
+			}
```
Comment (ziccardi) [inline internal/kafka/internal/clusters/standalone_provider_test.go:1238]: See previous comment about making this more readable

File: internal/kafka/internal/clusters/provider_test.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			d := &DefaultProviderFactory{
+				providerContainer: tt.fields.providerContainer,
+			}
+			got, err := d.GetProvider(tt.args.providerType)
+			Expect(err!=nil).To(Equal(tt.wantErr))
+			if got != nil{
+			Expect(got).To(Equal(tt.want))
+			}
```
Comment (ziccardi) [inline internal/kafka/internal/clusters/provider_test.go:115]: What if `got` is `nil` but `tt.want != nil` ?
Wouldn't be better to just get rid of the `if` and leave just the `Expect` ?

File: internal/kafka/internal/clusters/standalone_provider_test.go [hunk 7]
```diff
@@ -1046,3 +1040,202 @@ func TestStandaloneProvider_buildKasFleetshardOperatorSubscription(t *testing.T)
...
+			},
+		},
+	}
+	RegisterTestingT(t)
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			provider := newStandaloneProvider(test.fields.connectionFactory, test.fields.dataplaneClusterConfig)
+			ok, err := provider.InstallStrimzi(test.args.clusterSpec)
+			Expect(err != nil).To(Equal(test.wantErr))
+			Expect(ok).To(Equal(true))
```
Comment (ziccardi) [inline internal/kafka/internal/clusters/standalone_provider_test.go:1074]:
```suggestion
			Expect(ok).To(BeTrue())
```

**Review by ziccardi — CHANGES_REQUESTED** [review]
Left just a couple of comments more

File: internal/kafka/internal/clusters/provider_test.go (hunk 6 above)
Comment (JasonPowr) [reply internal/kafka/internal/clusters/provider_test.go:115]: I ran into an issue where the test wasnt passing when I left the Expect there. The error it was giving me was Refusing to compare <nil> to <nil>. Wasn't sure how to get around it other than this, Have you any suggestions ?

Comment (JasonPowr) [reply internal/kafka/internal/clusters/provider_test.go]: `Expect(got).To(Or(BeNil(),Equal(tt.want)))` this seems to solve the issue but im still happy for other suggestions that may be better.

Comment (ziccardi) [reply internal/kafka/internal/clusters/provider_test.go]: I believe the issue is this one: https://github.com/onsi/gomega/issues/267
Probably then the best we can do is something like this:
```go
Expect(got==nil).To(Equal(tt.want==nil))
if got != nil {
  Expect(got).To(Equal(tt.want))
}

```

File: internal/kafka/internal/clusters/standalone_provider_test.go [hunk 8]
```diff
@@ -1046,3 +1040,202 @@ func TestStandaloneProvider_buildKasFleetshardOperatorSubscription(t *testing.T)
...
+			},
+		},
+	}
+	RegisterTestingT(t)
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			provider := newStandaloneProvider(test.fields.connectionFactory, test.fields.dataplaneClusterConfig)
+			ok, err := provider.InstallStrimzi(test.args.clusterSpec)
+			Expect(err != nil).To(Equal(test.wantErr))
+			Expect(ok).To(BeTrue())
```
Comment (ziccardi) [inline internal/kafka/internal/clusters/standalone_provider_test.go:1074]: if `err != nil` and `test.wantErr == true`, I believe ok won't be `true` and the test will fail.

File: internal/kafka/internal/clusters/standalone_provider_test.go [hunk 9]
```diff
@@ -1046,3 +1040,202 @@ func TestStandaloneProvider_buildKasFleetshardOperatorSubscription(t *testing.T)
...
+			},
+		},
+	}
+	RegisterTestingT(t)
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			provider := newStandaloneProvider(test.fields.connectionFactory, test.fields.dataplaneClusterConfig)
+			ok, err := provider.InstallKasFleetshard(test.args.clusterSpec, test.args.params)
+			Expect(err != nil).To(Equal(test.wantErr))
+			Expect(ok).To(BeTrue())
```
Comment (ziccardi) [inline internal/kafka/internal/clusters/standalone_provider_test.go:1118]: See previous comment

File: internal/kafka/internal/clusters/standalone_provider_test.go [hunk 10]
```diff
@@ -1046,3 +1040,202 @@ func TestStandaloneProvider_buildKasFleetshardOperatorSubscription(t *testing.T)
...
+		},
+	}
+
+	RegisterTestingT(t)
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			provider := newStandaloneProvider(test.fields.connectionFactory, test.fields.dataplaneClusterConfig)
+			ok, err := provider.AddIdentityProvider(test.args.clusterSpec, test.args.identityProvider)
+			Expect(err != nil).To(Equal(test.wantErr))
+			Expect(ok).To(Equal(test.want))
```
Comment (ziccardi) [inline internal/kafka/internal/clusters/standalone_provider_test.go:1173]: See previous comment

**Review by ziccardi — APPROVED** [review]
Awesome job!

**Pull request merged**